lib/puppet/pal/pal_impl.rb

```diff
@@ -440,6 +439,10 @@ def self.main(manifest, facts, variables, internal_compiler_class)
         # TRANSLATORS: Do not translate, symbolic name
         Puppet.override(overrides, "PAL::with_#{internal_compiler_class}_compiler") do
           compiler.compile do | compiler_yield |
+            # In case the varaibles passed to the compiler are PCore types defined in modules, they
+            # need to be deserialized and added from within the this scope, so that loaders are
+            # available during deserizlization.
+            add_variables(compiler.topscope, Puppet::Pops::Serialization::FromDataConverter.convert(pal_variables))
```

Comment:
Why cant we add give https://github.com/puppetlabs/puppet/blob/f29b9681e72271ca7a1dc52932d566eb217ace52/lib/puppet/pal/catalog_compiler.rb#L10 an add_variables method? Then we could just call that in bolt/puppetserver?

Reply:
We can, but it seems incorrect to me that PAL would provide parameters for setting variables and not have them work for anything other than core puppet datatypes.

If PAL supports providing variables as a parameter to `.with_catalog_compiler` it should support those variables if they are pcore types defined in modules too.